File: app/local_plugins/core/ai_plugin_interface/android/src/main/kotlin/com/aiagent/plugin_interface/NativeLlmService.kt
package com.aiagent.plugin_interface

/**
 * LLM（大语言模型）服务原生接口
 *
 * 实现方：llm_openai 等服务插件
 * 使用方：agent_chat 等 Agent 类型插件
 *
 * 职责：HTTP SSE 流式推理 + 工具调用 + 取消
 */
interface NativeLlmService {

    /**
     * 初始化 LLM 服务
     * @param configJson  服务配置 JSON（apiKey, baseUrl, model, temperature 等）
     */
    fun initialize(configJson: String)

    /**
     * 流式对话
     *
     * 挂起函数：直到推理完成才返回完整文本。
     * 通过 callback 推送流式增量。
     *
     * @param requestId  请求 ID
     * @param messages   消息历史 [{"role":"user","content":"..."},...]
     * @param tools      工具定义（MCP / Function Calling）
     * @param callback   流式回调
     * @return 完整的 assistant 回复文本
     */
    suspend fun chat(
        requestId: String,
        messages: List<Map<String, Any>>,
        tools: List<Map<String, Any>> = emptyList(),
        callback: LlmCallback,
    ): String

    /** 取消当前请求 */
    fun cancel()
}

/**
 * LLM 流式回调
 */
interface LlmCallback {
    /** 第一个 token 到达 */
    fun onFirstToken(textDelta: String)

    /** 后续 token 增量 */
    fun onTextDelta(textDelta: String)

    /** 思考增量（部分模型支持） */
    fun onThinkingDelta(delta: String)

    /** 工具调用开始 */
    fun onToolCallStart(id: String, name: String)

    /** 工具调用参数增量 */
    fun onToolCallArguments(delta: String)

    /** 工具调用结果 */
    fun onToolCallResult(result: String)

    /** 推理完成 */
    fun onDone(fullText: String)

    /** 错误 */
    fun onError(code: String, message: String)
}

File: apps/client/local_plugins/core/ai_plugin_interface/ios/Classes/NativeLlmService.swift
import Foundation

/// LLM (chat-style) service contract.
///
/// Implementations stream SSE responses, surface tool calls, and obey
/// per-request cancellation.
public protocol NativeLlmService: AnyObject {
    /// `configJson` carries apiKey / baseUrl / model / temperature / …
    func initialize(configJson: String)

    /// Stream a chat completion.
    ///
    /// - Parameter messages: history `[{role, content}, …]`.
    /// - Parameter tools: tool/function-calling definitions.
    /// - Parameter callback: streaming hooks; the final value is also returned.
    /// - Returns: the full assistant text (post-tool-call fold-in).
    func chat(
        requestId: String,
        messages: [[String: Any]],
        tools: [[String: Any]],
        callback: LlmCallback
    ) async throws -> String

    /// Best-effort cancel of the in-flight request.
    func cancel()
}

public extension NativeLlmService {
    func chat(
        requestId: String,
        messages: [[String: Any]],
        callback: LlmCallback
    ) async throws -> String {
        try await chat(requestId: requestId, messages: messages, tools: [], callback: callback)
    }
}

public protocol LlmCallback: AnyObject {
    /// First text token has arrived.
    func onFirstToken(textDelta: String)

    /// Subsequent text deltas.
    func onTextDelta(_ textDelta: String)

    /// Reasoning/thinking deltas (supported by some providers).
    func onThinkingDelta(_ delta: String)

    /// A tool call has been opened by the model.
    func onToolCallStart(id: String, name: String)

    /// Streamed arguments JSON delta for the current tool call.
    func onToolCallArguments(_ delta: String)

    /// Tool execution result; sent before the model resumes its reply.
    func onToolCallResult(_ result: String)

    /// Streaming finished cleanly.
    func onDone(fullText: String)

    /// Streaming failed.
    func onError(code: String, message: String)
}
